working/extension_structs/overview.md

```diff
@@ -0,0 +1,815 @@
+  implements Comparable<Component> {
+
+  double get x => _x/_w;
+  double get y => _x/_w;
+  double get z => _z/_w;
+
+  int compareTo(Component other) => throw "TODO: Unimplemented";
+
+  Component.zero() : _x = 0, _y = 0, _z = 0;
+}
+```
+
+With the addition of pattern matching and switches, structs will support closed
+families (algebraic datatypes).
+
+```dart
+abstract struct Operand;
+
+struct ConstantOperand(Value c) extends Operand;
+
+struct IdentifierOperand(Identifier i) extends Operand;
+
+Operand replaceIn(Operand oper, Identifier t, Value c) {
+  return switch (oper) {
+    case ConstantOperand(_) : oper;
+    case IdentifierOperand(var i) where i == t: ConstantOperand(c);
+    case IdentifierOperand(_) : oper;
+  }
+}
+```
+
+## Structs in more detail
+
+### Introduction form
+
+We add a new keyword "struct", which is used in place of "class".  The class
+name (plus generics if applicable) must be followed by a parenthesized list of
+field declarations (more on this below). There may optionally be an "extends"
+clause, and an "implements" list.  No mixins are permitted.  After the header,
+the usual brace delimited list of members may be provided subject to
+restrictions described below.  An empty set of members may be elided in favor of
+a semi-colon.  The simplest possible struct definitions then would look like:
+
+```dart
+struct Data();
+struct GenericData<T>();
+```
+
+#### Alternatives
+
+
+##### Re-use the class keyword
+
+An alternative is to continue to use "class" and to use some other piece of
+syntax to indicate that the object in question is not a general class.  For
+example, the parenthesized list might be enough:
+
+
+```dart
+class Data();
+class GenericData<T>();
+```
+
+Alternatively, an additional keyword could be used:
+
+```dart
+class Data wraps ();
+class GenericData<T> wraps ();
+```
+
+Other keywords could be considered.
+
+##### Choose a different leading keyword
+
+Instead of "struct", we could use "view", or some other choice:
+
+```dart
+view Data();
+view GenericData<T>();
+```
+
+##### Use a modifier on classes
+
+
+We could use "data class":
+
+```dart
+data class Data();
+data class GenericData<T>();
+```
+
+Another option would be "view class".
+
+
+### Fields and primary constructors
+
+The only new piece of syntax in this section of the proposal (other than the
+"struct" keyword) is the primary constructor which follows immediately after the
+class name + generics.  The primary constructor serves both to define the
+fields, and to define the signature of the default generated constructor.
+
+The primary constructor consists of a parenthesized list of comma separated
+`<type> <identifier> (= <expression>)?` entries.  That is, a list of variable
+declarations: with no modifiers; with types required; and with optional
+initializer values.
+
+Each entry in the list is a field in the struct.  Every field is implicitly
+final.  They may not be late, and a type must be provided.
```

Comment:
nit: consider mentioning this earlier in the proposal (it may be too buried in the details that indeed structs are immutable) :smile:

Reply:
done